### PR TITLE
Priorityq

### PR DESCRIPTION
**Pull request by Copenbacon**

### CODE REVIEW

File: src/pq.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+"""Implement a priority queue."""
+
+
+from math import inf
```
Comment (Bl41r) [inline src/pq.py]: This is not compatible in Python 2

File: src/pq.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class PikaQ(object):
```
Comment (Bl41r) [inline src/pq.py:6]: cute

File: src/pq.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                for val in vals:
+                    self.insert(val)
+            except ValueError:
+                print("Too many values inserted in one of your lists")
+        elif not hasattr(type(vals, "__iter__")):
+            self.insert(vals)
+        else:
+            raise ValueError("Need to submit either a single value, a list of values, or a list of values tupled/listed with priorities.")
+
+    def insert(self, item, pri=float(-inf)):
```
Comment (Bl41r) [inline src/pq.py]: after you remove the import statement, you will see that this is not quite correct

File: src/pq.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        self.pq[poppin_key].remove(poppin_list[0])
+        if len(poppin_list) == 0:
+            self.pq.pop(poppin_key)
+        return val_popped
+
+    def peek(self):
+        """Look at next item to be popped in the Priority Queue."""
+        if len(self.pq) > 0:
+            return str(min(self.pq)) + ": " + str(self.pq[min(self.pq)][0])
+        else:
```
Comment (Bl41r) [inline src/pq.py:64]: Why are you returning a string here talking about it being empty?  That is not the way to go.

Comment (Bl41r) [reply src/pq.py:64]: What would be more appropriate to return here?

File: src/pq.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        if type(item[1]) is str or hasattr(type(item[1]), "__iter__") or type(item[1]) is bool:
+            raise TypeError("Priority must be a number.")
+        if item[1] in self.pq:
+            self.pq[item[1]].append(item[0])
+        else:
+            self.pq[item[1]] = [item[0]]
+
+    def pop(self):
+        """Remove the highest priority value from the priority queue and return it."""
+        if len(self.pq) == 0:
```
Comment (Bl41r) [inline src/pq.py]: This works, assuming your len works, but why not just try/except this?  It is more pythonic

File: src/pq.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+"""Implement a priority queue."""
+import sys
+
+if sys.version_info[0] == 3:
```
Comment (Bl41r) [inline src/pq.py]: It is good accounting for different versions, but float('inf') works for both python2/3
it is the math.inf that does not work in both

Comment (Bl41r) [reply src/pq.py]: Also, what happens here when python4 comes out?  You can try/except stuff like this also (if you need to, which you don't here)

File: src/pq.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(vals=None): Initialize a priority queue, and insert any values passed in into the Priority Queue. Must pass a list of items.
+    def insert(item, pri=float(-inf)): Insert an item into the Priority Queue, if no priority specified, -inf will be the priority,
+    def pop(): Remove the highest priority value from the priority queue.
+    def peek(): Look at next item to be popped in the Priority Queue.
+
+
+    """
+
+    def __init__(self, vals=None):
+        """Initialize a priority queue."""
```
Comment (Bl41r) [inline src/pq.py:18]: Not sure if this is the best way to do this.  It seems a bit overly complex.

File: src/pq.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                except TypeError:
+                    for val in vals:
+                        self.insert(val)
+                except ValueError:
+                    raise ValueError("Submitting a string? Put it in a list or tuple first, and give it a priority! Otherwise it will be placed in the highest priority group.")
+
+    def insert(self, val, pri=float(-inf)):
+        """Insert an item into the Priority Queue according to its priority."""
+        item = (val, pri)
+        if type(item[1]) is str or hasattr(type(item[1]), "__iter__") or type(item[1]) is bool:
```
Comment (Bl41r) [inline src/pq.py:40]: This could probably be made to just check for an int or float

File: src/pq.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
+"""Implement a priority queue."""
+
+inf = float("inf")
```
Comment (nhuntwalker) [inline src/pq.py:3]: Global variables should be caps

File: src/pq.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    def __init__(vals=None): Initialize a priority queue, and insert any values passed in into the Priority Queue. Must pass a list of items.
+    def insert(item, pri=float(-inf)): Insert an item into the Priority Queue, if no priority specified, -inf will be the priority,
+    def pop(): Remove the highest priority value from the priority queue.
+    def peek(): Look at next item to be popped in the Priority Queue.
+
+
+    """
+
+    def __init__(self, vals=None):
```
Comment (nhuntwalker) [inline src/pq.py:17]: The spec didn't call for the constructor to take an iterable.

File: src/pq.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                try:
+                    for val, pri in vals:
+                        self.insert(val, pri)
+                except TypeError:
+                    for val in vals:
+                        self.insert(val)
+                except ValueError:
+                    raise ValueError("Submitting a string? Put it in a list or tuple first, and give it a priority! Otherwise it will be placed in the highest priority group.")
+
+    def insert(self, val, pri=float(-inf)):
```
Comment (nhuntwalker) [inline src/pq.py:37]: wrapping this in `float` is unnecessary. The `inf` variable that you set is already a float.

File: src/pq.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        """Insert an item into the Priority Queue according to its priority."""
+        item = (val, pri)
```
Comment (nhuntwalker) [inline src/pq.py:39]: Why did you need to make this its own separate item?

File: src/pq.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        if type(item[1]) is str or hasattr(type(item[1]), "__iter__") or type(item[1]) is bool:
+            raise TypeError("Priority must be a number.")
+        if item[1] in self.pq:
+            self.pq[item[1]].append(item[0])
+        else:
+            self.pq[item[1]] = [item[0]]
```
Comment (nhuntwalker) [inline src/pq.py:45]: These 4 lines could be come one by just using `setdefault`

File: src/pq.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        except ValueError:
+            raise IndexError("Empty PikaQ is un-poppable.")
+        if len(poppin_list) == 0:
+            self.pq.pop(poppin_key)
+        return val_popped
+
+    def peek(self):
+        """Look at next item to be popped in the Priority Queue."""
+        if len(self.pq) > 0:
+            return str(min(self.pq)) + ": " + str(self.pq[min(self.pq)][0])
```
Comment (nhuntwalker) [inline src/pq.py:63]: Why are you returning a string here? That's not what was called for by the specification.
